### ISSUE TITLE
Performing element-wise multiplication on adjacency matrix

### ISSUE THREAD

**Issue by zc-alexfan** [rusty1s/pytorch_geometric#541]
## ❓ Questions & Help

I am working on a "hard attention" version of GCN on this [paper](https://arxiv.org/abs/1904.03345). We know that in `GCNConv`, the new node features are updated by: 
~~~
X := g(WXA)
~~~
where A is the adjacency matrix, W is some learnable weights, g is an activation function. 

In contrast, the hard attention paper above has the updating rule as: 
~~~
X:= g(WX s(M dot A))
~~~
where `M` is a learnable weight matrix, `dot` is matrix element-wise multiplication and `s( . )` is a softmax function to normalize the adjacency matrix such that the attention weights of a node's incoming edges sum to 1. 

I am trying to implement this version of GCN (and perhaps create a pull request later), but I am confused about how to compute `M dot A` since `A` is represented by `edge_index` instead of an actual matrix.

**Comment by rusty1s** [comment]
IMO, the operator you mentioned is highly related to the implementation of the `GATConv`, so instead of learning a dense mask `M`, you learn a sparse mask `M` *for each existing edge*, so *e.g.*:
``` python
M = Parameter(torch.Tensor(num_nodes, num_nodes))
...
row, col = edge_index
sparse_M = M[row, col]
sparse_M = softmax(row, sparse_M, num_nodes)
```

**Comment by MilkshakeForReal** [comment]
I'm also working on it  and we could cooperate if u r interested!!
